source/io.c

```diff
@@ -342,24 +466,100 @@ napi_value aws_napi_io_tls_ctx_new(napi_env env, napi_callback_info info) {
+    if (!aws_napi_is_null_or_undefined(env, node_pkcs11_options)) {
+
+        napi_value node_pkcs11_lib = NULL;
+        AWS_NAPI_CALL(env, napi_get_named_property(env, node_pkcs11_options, "pkcs11_lib", &node_pkcs11_lib), {
+            napi_throw_type_error(env, NULL, "'pkcs11_lib' is required for PKCS#11");
+            goto cleanup;
+        });
+
+        napi_value node_pkcs11_lib_handle = NULL;
+        AWS_NAPI_CALL(env, napi_get_named_property(env, node_pkcs11_lib, "handle", &node_pkcs11_lib_handle), {
+            napi_throw_type_error(env, NULL, "'pkcs11_lib' must be a Pkcs11Lib");
+            goto cleanup;
+        });
+
+        struct pkcs11_lib_binding *pkcs11_lib_binding = NULL;
+        AWS_NAPI_CALL(env, napi_get_value_external(env, node_pkcs11_lib_handle, (void **)&pkcs11_lib_binding), {
+            napi_throw_type_error(env, NULL, "'pkcs11_lib' must be a Pkcs11Lib");
+            goto cleanup;
+        });
+        pkcs11_options.pkcs11_lib = pkcs11_lib_binding->native;
+
+        /* user_pin property is required. null is allowed, but this is unusual, so we require the user to set it */
+        napi_value node_user_pin = NULL;
+        AWS_NAPI_CALL(env, napi_get_named_property(env, node_pkcs11_options, "user_pin", &node_user_pin), {
+            napi_throw_type_error(env, NULL, "'user_pin' is required for PKCS#11 (must be string or null)");
+            goto cleanup;
+        });
+
+        napi_valuetype node_user_pin_type = napi_undefined;
+        napi_typeof(env, node_user_pin, &node_user_pin_type);
+        if (node_user_pin_type == napi_undefined) {
+            napi_throw_type_error(env, NULL, "user_pin' is required for PKCS#11 (must be string or null)");
+            goto cleanup;
+        }
+
+        if (node_user_pin_type != napi_null) {
+            if (napi_ok != aws_byte_buf_init_from_napi(&pkcs11_pin, env, node_user_pin)) {
+                napi_throw_type_error(env, NULL, "PKCS#11 'user_pin' must be a string or null");
+                goto cleanup;
+            }
+            pkcs11_options.user_pin = aws_byte_cursor_from_buf(&pkcs11_pin);
+        }
+
+        napi_value node_slot_id = NULL;
+        if (napi_ok == napi_get_named_property(env, node_pkcs11_options, "slot_id", &node_slot_id)) {
+            if (!aws_napi_is_null_or_undefined(env, node_slot_id)) {
+                if (napi_ok != napi_get_value_int64(env, node_slot_id, (int64_t *)&pkcs11_slot_id)) {
+                    napi_throw_type_error(env, NULL, "PKCS#11 'slot_id' must be an int");
+                    goto cleanup;
+                }
+                pkcs11_options.slot_id = &pkcs11_slot_id;
+            }
+        }
+
+        /* clang-format off */
+
+        #define PARSE_PKCS11_OPTION_STR(property_name, option_cursor, storage_buffer)                                  \
+        do {                                                                                                           \
```

Comment:
Can't we just use a scoped block here?

Reply:
it's the classic "put your macro in a `do{}while(0)` so users are forced to put semicolon after it" trick

but yeah, confusing, and this macro not a global helper

I'll simplify